Resource control based on cgroup (#7)
ray-project/enhancements: reps/2022-04-01-resource-control.md

## Summary
### General Motivation

Currently, we don't control and limit the resource usage of worker processes, except running the worker in a container (see the `container` part in the [doc](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference)). In most of the scenarios, the container is unnecessary, but resource control is necessary for isolation.

[Control groups](https://man7.org/linux/man-pages/man7/cgroups.7.html), usually referred to as cgroups, are a Linux kernel feature which allow processes to be organized into hierarchical groups whose usage of various types of resources can then be limited and monitored.

So, the goal of this proposal is to achieve resource control for worker processes by cgroup in Linux.

### Should this change be within `ray` or outside?

> These changes would be within Ray core.

Reviewer: I actually think this is a prime candidate for being a third-party runtime_env plugin (which we can either keep in a separate repo, or as a "contrib" module to Ray.

## Stewardship
### Required Reviewers
@ericl, @edoakes, @simon-mo, @chenk008 @raulchen 

### Shepherd of the Proposal (should be a senior committer)
@ericl

## Design and Architecture

### Cluster level API
We should add some new system configs for the resource control.

> - worker_resource_control_method: Set to "cgroup" by default.

Reviewer 2: cgroup is the only method at the moment right?

Author: Right. But we can also support others, e.g. a custom monitor which could kill worker processes when the resources exceeded.

- cgroup_manager: Set to "cgroupfs" by default. We should also support `systemd`.
- cgroup_mount_path: Set to `/sys/fs/cgroup/` by default.

> - cgroup_unified_hierarchy: Whether to use cgroup v2 or not. Set to `False` by default.

Reviewer 2: can we only support cgroup v2 to reduce the scope?

Author: I think we should also support cgroup v1 because it has been used widely and written into standard. I have described here https://github.com/ray-project/enhancements/pull/7/files#diff-98ccfc9582e95581aae234797bc273b2fb68cb9e4dcc3030c8e94ba447daef7dR112-R113.
And in Ant, we only support v1 in production. Can you check your production environments and CI environments?

- cgroup_use_cpuset: Whether to use cpuset. Set to `False` by default.

### User level API
#### Simple API using a flag
```python
    runtime_env = {

>         "enable_resource_control": True,

Reviewer: Is there any reason this cannot be implemented using runtime_env plugins? We should consider whether to extend the plugin mechanism to support this, if this is not possible, rather than special casing cgroups.

The goal of runtime env plugins is exactly to enable supporting these kind of complex use cases in a customizable way.

Reviewer 3: In the current runtime env plugin mechanism design, a plugin must implement a whole runtime env type (e.g., pip, conda, etc).
Cgroup is different here. Because it is more like a config of runtime env, rather than a new runtime env type. Because it can be used together with other types. 
So in order to pluginize cgroup, we may need a new plugin mechanism that can extend the the current plugin mechanism. This seems weird. and I'm not sure if such a new plugin mechanism can be re-used for other features. 
So I think maybe we can consider pluginizing it later if we have a second similar requirement.

Author: How to achieve the logic of deleting cgroups in plugin?

Author: Sorry, let me explain what I said above.

with the current plugin mechanism, it's hard to implement cgroup deletion. As I said here https://github.com/ray-project/enhancements/pull/7#discussion_r847002806, the cgroups are one-off. It couldn't be shared and reused. So, the logic should be that: Create cgroup -> start worker -> worker dies -> delete cgroup. For example, if the cgroup dir is `/sys/fs/cgroup/{worker_id}`, we can't reuse the reference counting logic to delete this dir because we don't provide the `worker_id` when we decrease the reference by the PRC from raylet to agent.

  - (**Step 7**) The Raylet starts the new worker process with the `runtime_env_context`.
  - (**Step 8**) The `setup_worker.py` setups the `resource_control` by `command_prefix` for the new worker process.

#### Cgroup Manager
The Cgroup Manager is used to create or delete cgroups, and bind worker processes to cgroups. We plan to integrate the Cgroup Manager in the Agent.

We should abstract the Cgroup Manager because we have more than one way to manager cgroups in linux. The two main ways are cgroupfs and systemd, which also used in the [container technology](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/).

##### cgroupfs
Using cgroupfs to manage cgroups is like:
```
mkdir /sys/fs/cgroup/{worker_id}
echo "200000 1000000" > /sys/fs/cgroup/foo/cpu.max
echo {pid} > /sys/fs/cgroup/foo/cgroup.procs
```
NOTE: This is an example based on cgroup v2. The commmand lines in cgroup v1 is defferent and incompatible.

We also could use the [libcgroup](https://github.com/libcgroup/libcgroup/blob/main/README) to simplify the implementation. This library support both cgroup v1 and cgroup v2. 

##### systemd

> Using systemd to manage cgroups is like:

Reviewer 2: what's the advantage of using systemd here?

Author: If we use systemfs in a systemd based system,  there will be more than one component which manage the cgroup tree. And from https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/, we could know that:

*In the short-term future writing directly to the control group tree from applications should still be OK, as long as the [Pax Control Groups](http://www.freedesktop.org/wiki/Software/systemd/PaxControlGroups/) document is followed. In the medium-term future it will still be supported to alter/read individual attributes of cgroups directly, but no longer to create/delete cgroups without using the systemd API. In the longer-term future altering/reading attributes will also be unavailable to userspace applications, unless done via systemd's APIs (either D-Bus based IPC APIs or shared library APIs for passive operations).*

In addition, systemd is highly recommended by `runc` for cgroup v2 https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md.

##### More references
- [Yarn NodeManagerCgroups](https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/NodeManagerCgroups.html)

#### About Cgroup v1 and v2
[Cgroup v2](https://www.kernel.org/doc/Documentation/cgroup-v2.txt) is more reasonable, but we should also support cgroup v1 because v1 is widely used and has been hard coded to the [OCI](https://opencontainers.org/) standards. You can see this [blog](https://www.redhat.com/sysadmin/fedora-31-control-group-v2) for more.

Change if cgroup v2 has been enabled in your linux system:
```
mount | grep '^cgroup' | awk '{print $1}' | uniq
```

And you can try to enable cgroup v2:
```
grubby --update-kernel=ALL --args="systemd.unified_cgroup_hierarchy=1"
reboot
```

#### How to delete cgroups
When worker processes die, we should delete the cgroup which is created for the processes. This work is only needed for cgroupfs because systemd could handle it.

> So, we should add a new RPC named `CleanForDeadWorker` in `RuntimeEnvService`. The Raylet should send this PRC to the Agent and the Agent will delete the cgroup.

Reviewer 2: can this be part of RuntimeEnv's refcounting and GC RPCs?

Author: Good question. I have thought about this before. Currently, the URIs could be shared and reused between different worker processes. But for cgroups, it could only be used for one process and it is one-off. So, the existing logic couldn't be used directly.

Reviewer 4: Why not just making URI not shareable? I believe it should be easy to implement? (e.g., add worker id to each group URI key? )

## Summary
### General Motivation

Currently, we don't control and limit the resource usage of worker processes, except running the worker in a container (see the `container` part in the [doc](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference)). In most of the scenarios, the container is unnecessary, but resource control is necessary for isolation.

[Control groups](https://man7.org/linux/man-pages/man7/cgroups.7.html), usually referred to as cgroups, are a Linux kernel feature which allow processes to be organized into hierarchical groups whose usage of various types of resources can then be limited and monitored.

So, the goal of this proposal is to achieve resource control for worker processes by cgroup in Linux.

### Should this change be within `ray` or outside?

The resource control ability will be implemented in a runtime environments plugin. So the changes should no longer be within Ray core, but scoped to ray.experimental.cgroup_env_plugin (moved to util eventually).

## Stewardship
### Required Reviewers
@ericl, @edoakes, @simon-mo, @chenk008 @raulchen 

### Shepherd of the Proposal (should be a senior committer)

> @ericl

Reviewer 5: ```suggestion
@edoakes
```

4. The enhancement of runtime environment pluggability

When worker processes die, we should delete the cgroup which is created for the processes. This work is only needed for cgroupfs because systemd can automatically delete the cgroup when the process dies.

So, we should make some changes to allow each runtime environment plugin to do some work when worker processes die.
- Add `worker_id` field to the existing PRC request `GetOrCreateRuntimeEnvRequest` and `DeleteRuntimeEnvIfPossibleRequest`. These two RPCs are used to create and delete runtime environments from the Raylet to the Agent.
- Add two methods, named `create_for_worker` and `delete_for_worker`, in the plugin interface. We already have two methods named `create` and `delete` in the interface. The difference is that, `delete` is called only when there is no worker using the runtime environment(The reference counting to be zero), but `delete_for_worker` is called every time when a process dies.

Another change is that we should support loading runtime env plugins when we start the Ray nodes. For example:
```
ray start --head --runtime-env-plugins="ray.experimental.cgroup_env_plugin:ray.experimental.test_env_plugin"

> ```

Reviewer 6: Is this different from `eager_install`, which already exists in Ray?

Author: Two things absolutely. This option is used to register plugins into the dashboard agent. For example, you can register a plugin named "conda" when start the ray node. So, when a job uses "conda" field  in `runtime_env` after the ray cluster starts, agent could call the "conda"  plugin directly to set up the environment.

Reviewer 6: Makes sense!  In the [pluggability design](https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit?usp=sharing) there's a related idea which is to use environment variables, e.g. `RAY_RUNTIME_ENV_PLUGIN_CGROUP=ray.experimental.test_cgroup_plugin ray start --head`.  These both seem straightforward to implement but I'm not sure which approach is the best, do you have any thoughts about this?

```

NOTE: The entire config options is [here](https://man7.org/linux/man-pages/man5/systemd.resource-control.5.html). And we can also use `StartTransientUnit` to create cgroup with worker process simply. This is a [dbus](https://www.freedesktop.org/wiki/Software/systemd/dbus/) API and there is a [dbus-python](https://dbus.freedesktop.org/doc/dbus-python/) module we can used.

#### Why we should support both cgroupfs and systemd
The cgroupfs and the systemd are two mainstream ways to manage cgroups. In [container technology](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/), this two ways are also supported. We should also support both of them.

The reason of using cgroupfs:
- Cgroupfs is a traditional and common way to manage cgroups. If your login user has the privilege of cgroup file system, you can create the cgroups.
- In some environments, you couldn't use systemd, e.g. in a non-systemd based system or in a container which couldn't access the systemd service.

The reason of using systemd:
- Systemd is highly recommended by [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md.) for cgroup v2.

> - If we use cgroupfs in a systemd based system, there will be more than one component which manage the cgroup tree.

Author: I think so in any cases.

> - In the systemd [docs](https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/), we can know that  create/delete cgroups will be unavailable to userspace applications, unless done via systemd's APIs.

Reviewer 3: what does this mean? what problem will occur if we use cgroupfs?

Author: This is a planing of systemd. It will occur in future.

- The systemd has a good abstract API of cgroups. 

> 

Reviewer 3: So the conclusion is that, we support both cgroupfs and systemd. And systemd will be used when it is available, otherwise we fallback to cgroupfs. Right? can you also add the conclusion to the doc?

Author: Maybe we should make it to be a config?

Reviewer 2: Can you clarify the design here? Is it fallback based or config based? If config based, what is the documentation for user here?

Author: @simon-mo Yep, I will add this details.

> ### Support using cgroups in Ray

Reviewer 3: ```suggestion
### Changes needed in Ray
```

1. Cgroup manager in the Agent

The Cgroup Manager is used to create or delete cgroups, and bind worker processes to cgroups. We plan to integrate the Cgroup Manager in the Agent. 

> We should abstract the Cgroup Manager because there are more than one way to manager cgroups in linux: cgroupfs and systemd.

Reviewer 3: ```suggestion
Cgroup Manager should expose an abstract interface that can hide the differences between cgroupfs/systemd and cgroup v1/v2.
```

## Summary
### General Motivation

Currently, we don't control and limit the resource usage of worker processes, except running the worker in a container (see the `container` part in the [doc](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#api-reference)). In most of the scenarios, the container is unnecessary, but resource control is necessary for isolation.

[Control groups](https://man7.org/linux/man-pages/man7/cgroups.7.html), usually referred to as cgroups, are a Linux kernel feature which allow processes to be organized into hierarchical groups whose usage of various types of resources can then be limited and monitored.

So, the goal of this proposal is to achieve resource control for worker processes by cgroup in Linux.

### Should this change be within `ray` or outside?

> These changes would be within Ray Core.

Reviewer: The changes should no longer be within Ray core, but scoped to `ray.experimental.cgroup_env_plugin` (moved to util eventually).

## Stewardship
### Required Reviewers
@ericl, @edoakes, @simon-mo, @chenk008 @raulchen 

### Shepherd of the Proposal (should be a senior committer)
@ericl

## Design and Architecture

### Cluster level API

> We should add some new system configs for the resource control.

Reviewer: There should be no system level configs for a runtime_env plugin.

- worker_resource_control_method: Set to "cgroup" by default.
- cgroup_manager: Set to "cgroupfs" by default. We should also support `systemd`.
- cgroup_mount_path: Set to `/sys/fs/cgroup/` by default.
- cgroup_unified_hierarchy: Whether to use cgroup v2 or not. Set to `False` by default.
- cgroup_use_cpuset: Whether to use cpuset. Set to `False` by default.

### User level API
Users could turn on/off resource control by setting the relevant fields of runtime env (in job level or task/actor level).
#### Simple API using a flag
```python
    runtime_env = {
        "enable_resource_control": True,
    }
```

```python
    from ray.runtime_env import RuntimeEnv
    runtime_env = ray.runtime_env.RuntimeEnv(
        enable_resource_control=True
    )
```

#### Entire APIs
```python
    runtime_env = {
        "enable_resource_control": True,

>         "resource_control_config": {

Reviewer 4: Is cgroup version automatically detected?

Author: I found a util function in runc https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L34-L50. But cgroup v1 and v2 could both be enabled in some systems. And it also depends on which sub systems(e.g. cpu, memory) has been enable in the cgroup. So, I'd like to add a config in first step. We can also change this in future.

            "cpu_enabled": True,
            "memory_enabled": True,
            "cpu_strict_usage": False,

>             "memory_strict_usage": True,

Reviewer 6: What does each of these options do?  Where would the memory limit and CPU limit numbers be specified in this API?  Sorry if these are well-known options, I'm still learning about cgroups.

Author: I will add some description for each of the options. We already support setting CPU number in the remote call option, for example:
```
@ray.remote(num_cpus=1)
class actor:
    ...
```

        }
    }
```

```python

>     from ray.runtime_env import RuntimeEnv, ResourceControlConfig

Reviewer: The config example should be rewritten to show what it looks like as a runtime_env plugin.

The reason of using cgroupfs:
- Cgroupfs is a traditional and common way to manage cgroups. If your login user has the privilege of cgroup file system, you can create the cgroups.
- In some environments, you couldn't use systemd, e.g. in a non-systemd based system or in a container which couldn't access the systemd service.

The reason of using systemd:
- Systemd is highly recommended by [runc](https://github.com/opencontainers/runc/blob/main/docs/cgroup-v2.md.) for cgroup v2.
- If we use cgroupfs in a systemd based system, there will be more than one components which simultaneously manage the cgroup tree.
- In the systemd [docs](https://www.freedesktop.org/wiki/Software/systemd/ControlGroupInterface/), we can know that, in future, create/delete cgroups will be unavailable to userspace applications, unless done via systemd's APIs.
- The systemd has a good abstract API of cgroups. 

So, we should support both cgroupfs and systemd. And we can provide a config which is used to change the cgroup manager.

> ### Changes needed in Ray

Reviewer: Please rewrite this to reflect the plugin architecture.
